projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/impala_error_handler.py

```diff
@@ -441,3 +444,76 @@ def _is_pool_error(exception: Exception) -> bool:
             classname_with_package=".*OperationalError.*",
             exception_message_matcher_regex=".*admission for timeout ms in pool.*",
         )
+
+    def _handle_memory_error(self, exception, recovery_cursor):
+        if errors.exception_matches(
+            exception,
+            classname_with_package="impala.error.OperationalError",
+            exception_message_matcher_regex=".*Memory limit exceeded:.*",
+        ):
+            # We are going to try to increase the memory limits and see if the query passes
+            # But we won't do anything if the sql statement itself sets a memory limit
+            self._log.info(
+                "Query failed with memory error. We are going to increase the memory limit."
+                f"Error was: {exception.__class__}: {str(exception)}"
+            )
+            if (
+                "set memory_limit="
+                in str(
+                    recovery_cursor.get_managed_operation().get_operation_parameters_tuple()
+                ).lower()
+            ):
+                self._log.info(
+                    "The SQL statement you are trying to execute contains a set memory_limit option "
+                    "so we are not going to handle the memory error."
+                )
+                return False
+
+            # Incrementally increase the memory limit and as a last resort try to set the memory to an extreme value
+            if recovery_cursor.get_retries() > 3:
+                #  We won't be able to handle this error by increasing limits
+                return False
+            if recovery_cursor.get_retries() > 2:
+                # An extreme case but let's try it as a last resort
+                recovery_cursor.execute("set memory_limit=512GB;")
+            elif recovery_cursor.get_retries() > 1:
+                self._update_memory_limit(exception, recovery_cursor, 2.0)
```

Comment:
Are the multipliers based on the current queries that we have internally?

Reply:
These multipliers were discussed during the testing with the Infra team and we agreed that they should work best. I don't think there's a "perfect" multiplier here.

Reply:
Why do they work best? What is the logic ?